common/json.h

```diff
@@ -90,13 +90,17 @@ enum JSONType { JSONType_Null, JSONType_String, JSONType_Bool, JSONType_Number,
 class JSONValue {
 	friend class JSON;
 
+private:
+	// Not implemented, and should not be used.
+	JSONValue(long long int numberValue);
```

Comment:
If the implementation is missing, won't the compiler complain?

Reply:
No. Only the linker, in case it is actually used (can only be used by JSONValue itself, since it is private).

Reply:
This should indeed not be an issue for the compiler, and I have not seen any linker complain about non-implemented functions when they are not used for a while. However I remember linkers on some platforms throwing an error about this a while ago. I am not sure if I misremember, if the standard has changed, or if this is simply an undefined behaviour and up to the linker to give an error or not (and most do not). It might be good to clarify that this is valid code with a defined behaviour before me merge this.

Reply:
It is a typical pattern used for non-copyable classes. Copy ctor and copy assignment operator are declared as private and not implemented. I never had a problem with this even with very old toolchains.

Maybe what you saw was link error on unimplemented *virtual* function? Because virtual functions must be resolved at link time.

Reply:
Right. I have seen a lot of copy constructor or assignment operators declared as private and with an empty implementation (or deleted in code using c++11), but I don't see non-implemented ones as often so I forgot. This is indeed perfectly valid (and we actually do this in ScummVM with the `NonCopyable` class).

You may be right that the errors I saw in the past were for virtual functions. My memory is a bit hazy on that as that was around 20 years ago for platforms that no longer exist (and I would also not necessarily trust the vendor compilers and linkers for this platforms - I still have nightmares about trying to write and compile template code with the SGI compiler on IRIX).

Reply:
:)

Reply:
By the way, the linker error for a missing implementation of virtual functions is rather cryptic:

```c++
struct Foo { virtual void func(); };
int main() { Foo foo; return 0; }
```

This is ld:
```
undefined reference to `vtable for Foo'
```

lld actually has a better description:
```
ld.lld: error: undefined symbol: vtable for Foo
>>> referenced by example.cpp:1 (<source>:1)
>>>               /tmp/example-6ae871.o:(Foo::Foo())
>>> the vtable symbol may be undefined because the class is missing its key function (see https://lld.llvm.org/missingkeyfunction)
```